File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/data/type/ValueTypes.ts
module api.data.type {

    /*
     * Types need to be named as in ValueTypes.java
     */
    export class ValueTypes {

        static DATA = new DataValueType();

        static STRING = new StringValueType();

        static HTML_PART = new HtmlPartValueType();

        static XML = new XmlValueType();

        static LOCAL_DATE = new LocalDateValueType();

        static LOCAL_TIME = new LocalTimeValueType();

        static LOCAL_DATE_TIME = new LocalDateTimeValueType();

        static DATE_TIME = new DateTimeValueType();

        static CONTENT_ID = new ContentIdValueType();

        static LONG = new LongValueType();

        static BOOLEAN = new BooleanValueType();

        static DOUBLE = new DoubleValueType();

        static GEO_POINT = new GeoPointValueType();

        static NODE_ID = new ValueType("NodeId");

        static ALL: ValueType[] = [
            ValueTypes.DATA,
            ValueTypes.STRING,
            ValueTypes.HTML_PART,
            ValueTypes.XML,
            ValueTypes.LOCAL_DATE,
            ValueTypes.LOCAL_TIME,
            ValueTypes.LOCAL_DATE_TIME,
            ValueTypes.DATE_TIME,
            ValueTypes.CONTENT_ID,
            ValueTypes.LONG,
            ValueTypes.BOOLEAN,
            ValueTypes.DOUBLE,
            ValueTypes.GEO_POINT,
            ValueTypes.NODE_ID];

        public static fromName(name: string): ValueType {
            for (var i = 0; i < ValueTypes.ALL.length; i++) {
                var type = ValueTypes.ALL[i];
                if (type.toString() == name) {
                    return type;
                }
            }
            throw("Unknown ValueType: " + name);
        }
    }
}

File: modules/core/core-api/src/main/java/com/enonic/xp/data/ValueTypes.java
package com.enonic.xp.data;

import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.HashMap;
import java.util.Map;

import com.google.common.base.Preconditions;

import com.enonic.xp.annotation.PublicApi;
import com.enonic.xp.util.BinaryReference;
import com.enonic.xp.util.GeoPoint;
import com.enonic.xp.util.Link;
import com.enonic.xp.util.Reference;

@PublicApi
public final class ValueTypes
{
    public static final ValueType<PropertySet> PROPERTY_SET = new PropertySetValueType();

    public static final ValueType<String> STRING = new StringValueType();

    public static final ValueType<String> XML = new XmlValueType();

    public static final ValueType<LocalDate> LOCAL_DATE = new LocalDateValueType();

    public static final ValueType<LocalDateTime> LOCAL_DATE_TIME = new LocalDateTimeValueType();

    public static final ValueType<LocalTime> LOCAL_TIME = new LocalTimeValueType();

    public static final ValueType<Instant> DATE_TIME = new DateTimeValueType();

    public static final ValueType<Long> LONG = new LongValueType();

    public static final ValueType<Double> DOUBLE = new DoubleValueType();

    public static final ValueType<GeoPoint> GEO_POINT = new GeoPointValueType();

    public static final ValueType<Reference> REFERENCE = new ReferenceValueType();

    public static final ValueType<Link> LINK = new LinkValueType();

    public static final ValueType<Boolean> BOOLEAN = new BooleanValueType();

    public static final ValueType<BinaryReference> BINARY_REFERENCE = new BinaryReferenceValueType();

    private static final Map<String, ValueType> TYPES_BY_NAME = new HashMap<>();

    static
    {
        register( PROPERTY_SET );
        register( STRING );
        register( XML );
        register( LOCAL_DATE );
        register( LOCAL_TIME );
        register( LOCAL_DATE_TIME );
        register( DATE_TIME );
        register( LONG );
        register( DOUBLE );
        register( GEO_POINT );
        register( BOOLEAN );
        register( REFERENCE );
        register( LINK );
        register( BINARY_REFERENCE );
    }

    private static void register( ValueType valueType )
    {
        Object previous = TYPES_BY_NAME.put( valueType.getName(), valueType );
        Preconditions.checkState( previous == null, "ValueType already registered: " + valueType.getName() );
    }

    public static ValueType getByName( final String name )
    {
        return TYPES_BY_NAME.get( name );
    }
}
